Civi/Test/Api4TestTrait.php

```diff
@@ -0,0 +1,271 @@
+<?php
+
+namespace Civi\Test;
+
+use Civi\Api4\Generic\Result;
+use Civi\Api4\Utils\CoreUtil;
+
+/**
+ * Class Api3TestTrait
+ *
+ * @package Civi\Test
+ *
+ * This trait defines a number of helper functions for testing APIv4.
+ *
+ * This trait is intended for use with PHPUnit-based test cases.
+ */
+trait Api4TestTrait {
+
+  /**
+   * Inserts a test record, supplying all required values if not provided.
+   *
+   * Test records will be automatically deleted during tearDown.
+   *
+   * @param string $entityName
+   * @param array $values
+   * @return array|null
+   * @throws \CRM_Core_Exception
+   * @throws \Civi\API\Exception\NotImplementedException
+   */
+  public function createTestRecord(string $entityName, array $values = []): ?array {
+    return $this->saveTestRecords($entityName, ['records' => [$values]])->single();
+  }
+
+  /**
+   * Saves one or more test records, supplying default values.
+   *
+   * Test records will be automatically deleted during tearDown.
+   *
+   * @param string $entityName
+   * @param array $saveParams
+   * @return \Civi\Api4\Generic\Result
+   * @throws \CRM_Core_Exception
+   * @throws \Civi\API\Exception\NotImplementedException
+   */
+  public function saveTestRecords(string $entityName, array $saveParams): Result {
+    $saveParams += [
+      'checkPermissions' => FALSE,
+      'defaults' => [],
+    ];
+    $idField = CoreUtil::getIdFieldName($entityName);
+    foreach ($saveParams['records'] as &$record) {
+      $record += $saveParams['defaults'];
+      if (empty($record[$idField])) {
+        $this->getRequiredValuesToCreate($entityName, $record);
+      }
+    }
+    // Unset for clarity as it leaks from the foreach & is a reference.
+    unset($record);
+    $saved = civicrm_api4($entityName, 'save', $saveParams);
+    foreach ($saved as $item) {
+      $this->testRecords[] = [$entityName, [[$idField, '=', $item[$idField]]]];
```

Comment:
Can this `$testRecords` property be extracted too?

Reply:
as in `setTestRecords` & `addTestRecord`?

Reply:
As in `public $testRecords;` or `protected $testRecords;`

Reply:
hmm yeah ok - I'll add the getters & setters too if I'm going that

Reply:
Actually - it can be private so I don't need them